File: backend/app/services/substrate_orientation.py
"""Supplier crystal-plane specifications; never reinterpret direction indices."""

import re

from app.services.v2_field_source import load_field_source


def normalize_crystal_plane(value: str, material: str = "", quartz_type: str = "") -> str:
    rules = load_field_source()["substrate_orientation"]
    if not isinstance(value, str) or len(value) > rules["max_length"]:
        raise ValueError("substrate_crystal_plane: invalid text")
    value = value.strip().replace("−", "-").replace("（", "(").replace("）", ")")
    if not value:
        return ""
    if material == "quartz" and quartz_type == "fused_silica" and value != "amorphous":
        raise ValueError("substrate_crystal_plane: fused silica is amorphous")
    if value in rules["statuses"]:
        if material == "sapphire_al2o3" and value in {"amorphous", "polycrystalline"}:
            raise ValueError("substrate_crystal_plane: sapphire is single crystalline")
        if (
            material == "quartz"
            and quartz_type == "single_crystal_quartz"
            and value in {"amorphous", "polycrystalline"}
        ):
            raise ValueError("substrate_crystal_plane: quartz type contradicts orientation state")
        return value
    if material == "quartz" and quartz_type != "single_crystal_quartz":
        raise ValueError("substrate_crystal_plane: specify single-crystal quartz first")
    if material == "sapphire_al2o3":
        name = re.sub(r"[\s-]", "", value.lower())
        for plane, letter in rules["sapphire_plane_names"].items():
            if any(name == f"{letter}{suffix}" for suffix in rules["sapphire_name_suffixes"]):
                return plane
    if value.startswith("(") and value.endswith(")"):
        value = value[1:-1].strip()
    if re.fullmatch(r"[+-]?\d+(?:\s+[+-]?\d+){2,3}", value):
        indices = [int(part) for part in value.split()]
    elif re.fullmatch(r"(?:-?\d){3,4}", value):
        indices = [int(part) for part in re.findall(r"-?\d", value)]
    else:
        raise ValueError("substrate_crystal_plane: use three or four integer plane indices")
    if len(indices) not in rules["index_counts"].get(material, [3, 4]):
        raise ValueError("substrate_crystal_plane: index count does not match substrate")
    if (
        not all(abs(index) <= 2**53 - 1 for index in indices)
        or not any(indices)
        or (len(indices) == 4 and sum(indices[:3]) != 0)
    ):
        raise ValueError("substrate_crystal_plane: invalid zero plane or h+k+i is not zero")
    return "(" + " ".join(map(str, indices)) + ")"


def normalize_substrate_orientation(data: dict) -> dict:
    """Normalize new fields only; retain old combined records verbatim."""
    material = data.get("substrate_material") or ""
    quartz_type = data.get("quartz_type") or ""
    raw = data.get("substrate_crystal_plane")
    plane = normalize_crystal_plane(raw if raw is not None else "", material, quartz_type)
    if material == "quartz" and quartz_type == "fused_silica":
        plane = "amorphous"
    cut = data.get("substrate_cut_spec")
    if cut is not None and (
        not isinstance(cut, str)
        or not cut.strip()
        or len(cut) > load_field_source()["substrate_orientation"]["max_length"]
    ):
        raise ValueError("substrate_cut_spec: invalid supplier specification")
    if (plane == "supplier_cut") != bool(cut):
        raise ValueError("substrate_cut_spec: required only for a supplier cut")
    if plane == "supplier_cut" and material == "quartz" and quartz_type != "single_crystal_quartz":
        raise ValueError("substrate_crystal_plane: specify single-crystal quartz first")
    return {
        "substrate_crystal_plane": plane or None,
        "substrate_cut_spec": cut.strip() if cut else None,
    }

File: frontend-next/src/shared/substrate-orientation.ts
import { substrateOrientation } from '@/shared/generated/field-metadata'
import { canonicalOption, localizedOption } from '@/shared/field-i18n'
import i18n from '@/shared/i18n'

export function normalizeCrystalPlane(
  raw: string,
  material = '',
  quartzType = '',
): string {
  const invalid = () => {
    throw new Error('substrate_crystal_plane')
  }
  if (raw.length > substrateOrientation.max_length) invalid()
  let value = raw
    .trim()
    .replaceAll('−', '-')
    .replaceAll('（', '(')
    .replaceAll('）', ')')
  if (!value) return ''
  if (
    material === 'quartz' &&
    quartzType === 'fused_silica' &&
    value !== 'amorphous'
  )
    invalid()
  if (substrateOrientation.statuses.includes(value)) {
    if (
      material === 'sapphire_al2o3' &&
      ['amorphous', 'polycrystalline'].includes(value)
    )
      invalid()
    if (
      material === 'quartz' &&
      quartzType === 'single_crystal_quartz' &&
      ['amorphous', 'polycrystalline'].includes(value)
    )
      invalid()
    return value
  }
  if (material === 'quartz' && quartzType !== 'single_crystal_quartz') invalid()
  if (material === 'sapphire_al2o3') {
    const name = value.toLowerCase().replace(/[\s-]/g, '')
    const plane = Object.entries(
      substrateOrientation.sapphire_plane_names,
    ).find(([, letter]) =>
      substrateOrientation.sapphire_name_suffixes.some(
        (suffix) => `${letter}${suffix}` === name,
      ),
    )
    if (plane) return plane[0]
  }
  if (value.startsWith('(') && value.endsWith(')'))
    value = value.slice(1, -1).trim()
  const indices = /^[+-]?\d+(?:\s+[+-]?\d+){2,3}$/.test(value)
    ? value.split(/\s+/).map(Number)
    : /^(?:-?\d){3,4}$/.test(value)
      ? (value.match(/-?\d/g) ?? []).map(Number)
      : []
  if (
    !(substrateOrientation.index_counts[material] ?? [3, 4]).includes(
      indices.length,
    ) ||
    !indices.every(Number.isSafeInteger) ||
    !indices.some((index) => index !== 0) ||
    (indices.length === 4 && indices[0] + indices[1] + indices[2] !== 0)
  )
    invalid()
  return `(${indices.join(' ')})`
}

export function substratePlaneOptions(
  material: string,
  quartzType = '',
): string[] {
  if (material === 'quartz' && quartzType === 'fused_silica')
    return ['amorphous']
  if (material === 'quartz' && quartzType !== 'single_crystal_quartz')
    return ['not_provided']
  return [
    ...(substrateOrientation.presets[material] ?? []),
    ...substrateOrientation.statuses.filter(
      (status) =>
        !(
          material === 'sapphire_al2o3' ||
          (material === 'quartz' && quartzType === 'single_crystal_quartz')
        ) || !['amorphous', 'polycrystalline'].includes(status),
    ),
  ]
}

export function substratePlaneLabel(
  value: string,
  material: string,
  language: string,
): string {
  const letter =
    material === 'sapphire_al2o3'
      ? substrateOrientation.sapphire_plane_names[value]
      : null
  return letter
    ? i18n.t('entityLibrary.orientation.namedPlane', {
        letter,
        indices: value,
        lng: language,
      })
    : localizedOption(value, language)
}

export function substrateOrientationPayload(
  values: Record<string, unknown>,
): Record<string, string> {
  const material = canonicalOption(String(values.substrate_material ?? ''))
  const quartzType = canonicalOption(String(values.quartz_type ?? ''))
  const raw = String(values.substrate_crystal_plane ?? '')
  const normalized = normalizeCrystalPlane(raw, material, quartzType)
  const plane =
    material === 'quartz' && quartzType === 'fused_silica'
      ? 'amorphous'
      : normalized
  const cut = String(values.substrate_cut_spec ?? '').trim()
  if (
    (plane === 'supplier_cut') !== Boolean(cut) ||
    cut.length > substrateOrientation.max_length
  )
    throw new Error('substrate_cut_spec')
  if (
    plane === 'supplier_cut' &&
    material === 'quartz' &&
    quartzType !== 'single_crystal_quartz'
  )
    throw new Error('substrate_crystal_plane')
  return {
    ...(plane ? { substrate_crystal_plane: plane } : {}),
    ...(cut ? { substrate_cut_spec: cut } : {}),
  }
}

/** Editing a new version never rewrites the historical supplier text. */
export function legacySubstrateOrientation(
  source: Record<string, unknown>,
): Record<string, unknown> {
  if (
    source.substrate_crystal_plane != null ||
    source.substrate_polish != null ||
    !source.substrate_orientation_polish
  )
    return source
  const legacy = source.substrate_orientation_polish
  const raw =
    typeof legacy === 'object' ? (legacy as { value?: unknown }).value : legacy
  const polish =
    typeof legacy === 'object'
      ? (legacy as { option?: unknown }).option
      : undefined
  let plane = ''
  let cut: string | undefined
  if (raw) {
    try {
      plane = normalizeCrystalPlane(
        String(raw),
        canonicalOption(String(source.substrate_material ?? '')),
        String(source.quartz_type ?? ''),
      )
    } catch {
      plane = 'supplier_cut'
      cut = String(raw)
    }
  }
  return {
    ...source,
    substrate_crystal_plane: plane,
    substrate_cut_spec: cut,
    substrate_polish: polish,
  }
}
